Compiler crash on assertion failed:  cannot invokespecial ____, the interface is not a direct parent

## reproduction steps

The error can be reproduced with the following interfaces and classes:

`A.java`
```java
public interface A {
    default int foo() {
        return 0;
    }
}
```

`B.java`
```java
public interface B extends A {
    int bar();
}
```

`C.scala`
```scala
class C extends B {

  override def bar(): Int = {
    return 0
  }

  override def foo(): Int = {
    val f: () => Int = super.foo; f()
  }
}

object Test {
  def main(args: Array[String]): Unit = {
   var c = new C()
   c.foo()
  }
}
```

The compiler will generate the following error:

```
scalac: Error while emitting C.scala
assertion failed: 
  cannot invokespecial A.foo, the interface is not a direct parent.
     while compiling: /Users/jerry.peng/workspace/scala-test/src/C.scala
        during phase: jvm
     library version: version 2.12.15
    compiler version: version 2.12.15
  reconstructed args: -classpath /Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/charsets.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/ext/cldrdata.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/ext/dnsns.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/ext/jaccess.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/ext/localedata.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/ext/nashorn.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/ext/sunec.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/ext/sunjce_provider.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/ext/sunpkcs11.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/ext/zipfs.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/jce.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/jfr.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/jsse.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/management-agent.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/resources.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/rt.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/lib/dt.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/lib/jconsole.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/lib/sa-jdi.jar:/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/lib/tools.jar:/Users/jerry.peng/workspace/runtime/scala-test/out/production/scala-test:/Users/jerry.peng/.m2/repository/org/scala-lang/scala-reflect/2.12.15/scala-reflect-2.12.15.jar:/Users/jerry.peng/.m2/repository/org/scala-lang/scala-library/2.12.15/scala-library-2.12.15.jar -Xprint:jvm

  last tree to typer: Ident(f)
       tree position: line 10 of /Users/jerry.peng/workspace/scala-test/src/C.scala
            tree tpe: Function0
              symbol: value f
   symbol definition: val f: Function0 (a TermSymbol)
      symbol package: <empty>
       symbol owners: value f -> method foo -> class C
           call site: constructor Test in object Test in package <empty>

== Source file context for tree position ==

     7   }
     8 
     9   override def foo(): Int = {
    10     val f: () => Int = super.foo; f()
    11 //    val f: Int = C.super.foo(); f
    12   }
    13 }

```


**using Scala 2.12.15**

**using javac 1.8.0_292**

## problem

The code is valid and should be able to be compiled successfully.  Looking at the code generation from scalac compiler I see the issue if probably because of 

```
<superaccessor> <artifact> private def super$foo(): Int = C.super.foo()
```

calling 

```java
 C.super.foo()
```

is not valid in Java.  The code has to explicitly reference parent class to invoke the method i.e. `B.super.foo()`.

Adding a comment here: using a pure Java example:
```java
import java.util.function.*;

interface A {
    default int foo() {
        return 0;
    }
}

interface B extends A {
    int bar();
}

public class Foo implements B {
  @Override
  public int foo() {
    IntSupplier func = B.super::foo;
    return func.getAsInt();
  }

  @Override
  public int bar() { return 42; }
}
```
This code compiles fine with JDK8's javac, and the bytecode emitted for the lambda capture function is:
```
  private int lambda$foo$0();
    descriptor: ()I
    flags: ACC_PRIVATE, ACC_SYNTHETIC
    Code:
      stack=1, locals=1, args_size=1
         0: aload_0
         1: invokespecial #4                  // InterfaceMethod B.foo:()I
         4: ireturn
      LineNumberTable:
        line 16: 0
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0       5     0  this   LFoo;
```
Note that it's calling `invokespecial InterfaceMethod B.foo:()I` instead of `invokespecial InterfaceMethod A.foo:()I`.

Also note that in Java if the example is changed to `A.super::foo`, then javac complains with:
```
Foo.java:16: error: not an enclosing class: A
    IntSupplier func = A.super::foo;
                        ^
1 error
```

--- a/src/compiler/scala/tools/nsc/transform/Mixin.scala
+++ b/src/compiler/scala/tools/nsc/transform/Mixin.scala
@@ -243,6 +243,17 @@ abstract class Mixin extends Transform with ast.TreeDSL with AccessorSynthesis {
     }
   }
 
+  def registerRequiredDirectInterface(alias: Symbol, clazz: Symbol, msg: Type => String): Unit = {
+    val owner = alias.owner
+    if (owner.isJavaDefined && owner.isInterface) {
+      if (!clazz.parentSymbolsIterator.contains(owner)) {
+        val suggestedParent = exitingTyper(clazz.info.baseType(owner))
+        reporter.error(clazz.pos, msg(suggestedParent))
+      } else
+        erasure.requiredDirectInterfaces.getOrElseUpdate(clazz, mutable.Set.empty) += owner
+    }
+  }
+
   /** Add all members to be mixed in into a (non-trait-) class
    *  These are:
    *    for every mixin trait T that is not also inherited by the superclass:
@@ -346,13 +357,8 @@ abstract class Mixin extends Transform with ast.TreeDSL with AccessorSynthesis {
               reporter.error(clazz.pos, "Member %s of mixin %s is missing a concrete super implementation.".format(
                 mixinMember.alias, mixinClass))
             case alias1 =>
-              if (alias1.owner.isJavaDefined && alias1.owner.isInterface) {
-                if (!clazz.parentSymbolsIterator.contains(alias1.owner)) {
-                  val suggestedParent = exitingTyper(clazz.info.baseType(alias1.owner))
-                  reporter.error(clazz.pos, s"Unable to implement a super accessor required by trait ${mixinClass.name} unless $suggestedParent is directly extended by $clazz.")
-                } else
-                  erasure.requiredDirectInterfaces.getOrElseUpdate(clazz, mutable.Set.empty) += alias1.owner
-              }
+              registerRequiredDirectInterface(alias1, clazz, parent =>
+                s"Unable to implement a super accessor required by trait ${mixinClass.name} unless $parent is directly extended by $clazz.")
               superAccessor.asInstanceOf[TermSymbol] setAlias alias1
           }
         }
@@ -579,8 +585,11 @@ abstract class Mixin extends Transform with ast.TreeDSL with AccessorSynthesis {
          */
         def completeSuperAccessor(stat: Tree) = stat match {
           case DefDef(_, _, _, vparams :: Nil, _, EmptyTree) if stat.symbol.isSuperAccessor =>
-            debuglog(s"implementing super accessor in $clazz for ${stat.symbol} --> ${stat.symbol.alias.owner} . ${stat.symbol.alias}")
-            val body = atPos(stat.pos)(Apply(SuperSelect(clazz, stat.symbol.alias), vparams map (v => Ident(v.symbol))))
+            val alias = stat.symbol.alias
+            debuglog(s"implementing super accessor in $clazz for ${stat.symbol} --> ${alias.owner} . ${alias}")
+            registerRequiredDirectInterface(alias, clazz, parent =>
+              s"Unable to implement a super accessor, $parent needs to be directly extended by $clazz.")
+            val body = atPos(stat.pos)(Apply(SuperSelect(clazz, alias), vparams map (v => Ident(v.symbol))))
             val pt   = stat.symbol.tpe.resultType
 
             copyDefDef(stat)(rhs = enteringMixin(transform(localTyper.typed(body, pt))))

--- a/test/files/neg/t12523.check
+++ b/test/files/neg/t12523.check
@@ -1,0 +1,4 @@
+Test.scala:1: error: Unable to implement a super accessor, A needs to be directly extended by class C.
+class C extends B {
+      ^
+1 error

--- a/test/files/neg/t12523/A.java
+++ b/test/files/neg/t12523/A.java
@@ -1,0 +1,5 @@
+public interface A {
+    default int foo() {
+        return 41;
+    }
+}

--- a/test/files/neg/t12523/B.java
+++ b/test/files/neg/t12523/B.java
@@ -1,0 +1,3 @@
+public interface B extends A {
+    int bar();
+}

--- a/test/files/neg/t12523/Test.scala
+++ b/test/files/neg/t12523/Test.scala
@@ -1,0 +1,15 @@
+class C extends B {
+  override def bar(): Int = 1
+
+  override def foo(): Int = {
+    val f: () => Int = super.foo
+    f()
+  }
+}
+
+object Test {
+  def main(args: Array[String]): Unit = {
+   var c = new C()
+   assert(c.foo() + c.bar() == 42)
+  }
+}

--- a/test/files/run/t12523/A.java
+++ b/test/files/run/t12523/A.java
@@ -1,0 +1,5 @@
+public interface A {
+    default int foo() {
+        return 41;
+    }
+}

--- a/test/files/run/t12523/B.java
+++ b/test/files/run/t12523/B.java
@@ -1,0 +1,3 @@
+public interface B extends A {
+    int bar();
+}

--- a/test/files/run/t12523/Test.scala
+++ b/test/files/run/t12523/Test.scala
@@ -1,0 +1,15 @@
+class C extends B with A {
+  override def bar(): Int = 1
+
+  override def foo(): Int = {
+    val f: () => Int = super.foo
+    f()
+  }
+}
+
+object Test {
+  def main(args: Array[String]): Unit = {
+   var c = new C()
+   assert(c.foo() + c.bar() == 42)
+  }
+}

